### DISCUSSION TITLE
Accessing model record from a block

### DISCUSSION THREAD

**Discussion by agentphoenix** [awcodes/scribble#12] [General]
I didn't notice anything in the docs about (but please point me to the right spot if it is there), so was just curious if there's a way to get the model record where the content data is stored to the individual block classes? Would be awesome (especially with file uploads) to be able to do things with the underlying model record. Thanks!

**Comment by awcodes** [comment]
hmm. yea, this has been discussed with the tiptap editor plugin.

I'm not fully sure of the application though. Should a block be able to affect the $record since the block is relative to the fields state and not the record as a whole?

Can you give me a more concrete usage example so I can understand it better?

**Reply by agentphoenix** [reply]
I'm building a page builder now (and a form builder after that) using Scribble.

For the page builder, having access to `$page` would allow me to organize file uploads (like a background image or an image in a specific spot on the page) under a directory with the page ID (or something similar). For my use-cases, I wouldn't be modifying the record, so even if I was able to pass primitive values like an integer that would be a big help too.

I haven't thought through the form builder entirely yet, but at some point I'm going to want to fill the form with outside data (like a user info form), so being able to pass something in will be key to making sure I can retrieve the necessary information and display it back in the form if that makes sense.

**Reply by awcodes** [reply]
Wouldn't that break in the 'create' context when there isn't a record, yet?

**Reply by agentphoenix** [reply]
For my use case, it doesn't apply. You always create the page / form first and then go to a separate design page to build the page / form. I realize that doesn't necessarily help from a package perspective, but it isn't a concern for my use cases.

**Reply by awcodes** [reply]
Fair, but I have to worry about it. Lol.

**Reply by awcodes** [reply]
I'm also not sure how to totally accommodate this with the current set up for modals. If there wasn't an issue in core filament with mounted form component actions it would be fairly easy.

**Reply by agentphoenix** [reply]
What's the issue in core Filament with mounted form component actions?

**Reply by awcodes** [reply]
Technically, it's not an 'issue' in core, it's just the nature of how form component actions work in core.

For example,

Let's say you have a modal launched from a form component action. Inside that modal you have an editor field. You add some content to the field, then you want to make a link. When you click the link, (if that were using a form component action to open it's settings modal) then the original modal gets removed from the dom and the link settings modal opens. So, because the original modal was removed from the dom, any events fired from the submitting the link modal form can't be heard from the editor instance because it is no longer in the dom. So, the event gets fired, livewire reloads the previous modal with the editor instance, but it's too late at that point for the editor to reinitialize and hear the event. Also the cursor position in the editor would be lost since it was destroyed and reinitialized.

Hope that makes sense. LOL.

**Reply by awcodes** [reply]
If there was a way to Stack modals in core, it would work. That's why I made the Pounce plugin, but I don't like having to depend on it. Nor do I fully like it's implementation, but it works.

**Reply by agentphoenix** [reply]
Yeah that makes sense. Filament v4 perhaps? I can bug Dan about it for sure.

In the interim, is there any way of passing primitives in to the blocks from the `ScribbleEditor`?

```php
ScribbleEditor::make('blocks')
  ->profile(PageBuilderProfile::class)
  ->blockData(pageId: $this->model->id);
```

Even if it was something like that, I could at least get what I needed into the blocks to use for say setting the name of a folder for file uploads or filling data in a rendered form block. I started looking through the code last night, but didn't immediately see a place where I could do that.

**Reply by awcodes** [reply]
I'm honestly not sure of a way to achieve it at the moment, since the block modals are detached from the tool calling them and are separate livewire components.

Certainly open to suggests.

**Comment by agentphoenix** [comment]
My solution was to stick the model ID in the cache when the Livewire component mounts. I then grab the ID out of the cache from the custom blocks to be able to store uploads in folders with their ID.

I'd love to see something native to the package to handle this, but I understand it isn't a simple one to solve and relies on changes in Filament as well.
